addon/resolver.js

```diff
@@ -0,0 +1,55 @@
+const Resolver = require('ember-resolver')['default']
+
+Resolver.reopen({
+  expandLocalLookup: function (targetFullName, sourceFullName) { // eslint-disable-line complexity
+    const modulePrefix = this.namespace.modulePrefix
+    const podModulePrefix = this.namespace.podModulePrefix
+    const parsedTarget = this.parseName(targetFullName)
+    const parsedSource = this.parseName(sourceFullName)
+
+    let sourceName = parsedSource.fullNameWithoutType
+    let targetName = parsedTarget.fullNameWithoutType
+    const targetType = parsedTarget.type
+
+    // Local lookup only applies to targets referenced from a template
+    if (parsedSource.type !== 'template') {
+      return `${parsedTarget.type}:${sourceName}/${targetName}`
```

Comment:
Technically the concept of local lookup is not restricted to templates/components/partials. That just happens to be our current motivating use case.  I could easily see `expandLocalLookup` happening from a `route` for example (could simulate in tests via `getOwner(this).lookup('foo:bar-baz', { source: 'current/route/module' });`).  

I think for now, it may make sense to change this to a `throw` with a message suggesting to open an issue here with your use case 😈